extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/methods/ListMethodImplementor.java

```diff
@@ -29,8 +42,19 @@
      *         rel = "list",
      *         entityClassName = "com.example.Entity"
      *     )
-     *     public List list() {
-     *         return Entity.listAll();
+     *     public Response list() {
+     *         if (this.isPaged()) {
+     *            Page page = ...; // Extract page index and size from a UriInfo field and create a page instance.
```

Comment:
Why not via query params?

Reply:
It takes the values out of the query parameters but through `UriInfo`. I didn't want to add them as method attributes for two reasons. Firstly, pagination could be disabled and the method attributes defined in the interface would not make much sense. Secondly, having a shorter method definition will make it easier for the user to override it when modifying its properties with `@MethodProperties` annotation.

Reply:
Good points, however if the user is overriding it with a custom implementation (in the future when we support this) he will have a hard time getting to those paging parameters if he needs to. Because the `UriInfo` is not a `@Context` parameter, so field injection will be required.

TBH it's not entirely clear which approach is better long-term. I'll note however that you've switched to the less-useful `Reponse` return type rather than set response headers via something that doesn't leak to the signature, contrary to what you did with query parameters ;)

I don't have a strong answer here, so perhaps it's one of those situations where we could either ask for more opinions, such as @geoand or @loicmathieu or just flip a coin and wait for user feedback.

Reply:
_Each time someone return a JAX-RS `Response` a puppy dies ..._

As typical use case is to not override any methods so it will not be too much visible. But I like my method to return meaningful objects ... And JAX-RS response is not even a generic type :(

> if the user is overriding it with a custom implementation (in the future when we support this) he will have a hard time getting to those paging parameters if he needs to. Because the UriInfo is not a @Context parameter, so field injection will be required.

+1

Reply:
Is there a way to add the response headers without returning a `Response` instance? Or do you suggest to implement an interceptor?

Reply:
> Good points, however if the user is overriding it with a custom implementation (in the future when we support this) he will have a hard time getting to those paging parameters if he needs to. Because the `UriInfo` is not a `@Context` parameter, so field injection will be required.
> 
> TBH it's not entirely clear which approach is better long-term. I'll note however that you've switched to the less-useful `Reponse` return type rather than set response headers via something that doesn't leak to the signature, contrary to what you did with query parameters ;)
> 
> I don't have a strong answer here, so perhaps it's one of those situations where we could either ask for more opinions, such as @geoand or @loicmathieu or just flip a coin and wait for user feedback.

I am personally in favor of what @gytis did here with `UriInfo` as a field instead of adding the parameters to the method signature. When we have the override capability, then we'll certainly have to document the way for the user to retrieve the headers.

As for the use of `Response`, I agree it should be avoided if possible.

Reply:
> Is there a way to add the response headers without returning a Response instance? Or do you suggest to implement an interceptor?

There is only the RESTEasy `HttpResponse` context object, I'm afraid.

Reply:
So in that case JAX-RS `Response` instance is better, no? At least it's from a spec.